hazelcast/src/main/java/com/hazelcast/config/ClassFilter.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static java.util.Collections.unmodifiableSet;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Holds blacklist and whitelist configuration in java deserialization configuration.
+ */
+public class ClassFilter {
+
+    private final Set<String> classes = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>());
+    private final Set<String> packages = Collections.newSetFromMap(new ConcurrentHashMap<String, Boolean>());
+
+    /**
+     * Returns unmodifiable set of class names.
+     */
+    public Set<String> getClasses() {
+        return unmodifiableSet(classes);
+    }
+
+    /**
+     * Returns unmodifiable set of package names.
+     */
+    public Set<String> getPackages() {
+        return unmodifiableSet(packages);
+    }
+
+    public ClassFilter addClasses(String... names) {
+        checkNotNull(names);
+        for (String name : names) {
+            classes.add(name);
+        }
+        return this;
+    }
+
+    public ClassFilter setClasses(Collection<String> names) {
+        checkNotNull(names);
+        classes.clear();
+        classes.addAll(names);
+        return this;
+    }
+
+    public ClassFilter addPackages(String... names) {
+        checkNotNull(names);
+        for (String name : names) {
+            packages.add(name);
+        }
+        return this;
+    }
+
+    public ClassFilter setPackages(Collection<String> names) {
+        checkNotNull(names);
+        packages.clear();
+        packages.addAll(names);
+        return this;
+    }
+
+    public boolean isEmpty() {
+        return classes.isEmpty() && packages.isEmpty();
+    }
+
+    public boolean isListed(String className) {
+        if (classes.contains(className)) {
+            return true;
+        }
+        if (!packages.isEmpty()) {
+            int dotPosition = className.lastIndexOf(".");
+            if (dotPosition > 0) {
+                String packageName = className.substring(0, dotPosition);
```

Comment:
This will lead to litter. You can optimize it by storing the clasname in the 'classes'. So this substring check will be done one per class and then it will be stored in 'classes'. Then you need to make classes modifiable by using either a CHM or replace the whole collection and consider it immutable.

This is the second time I placed this comment.

Reply:
This is the second time I object the suggestion for adding classes to a Set if their packages are listed. It would create another unsafe piece in the code - possible DOS by causing OOM.

Reply:
Hmm, can't we settle this issue with a JFR test? The current implementation vs. a cached one. And a normal usage scenario vs. a DOS based one (so a handful of classes vs. an attack based with thousands of generated classes). Would be four scenarios in total to compare.

Reply:
```
package com.hazelcast;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.Warmup;
import org.openjdk.jmh.infra.Blackhole;

import java.util.concurrent.TimeUnit;


@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Thread)
@Fork(value = 1, warmups = 1)
@Warmup(iterations = 2)
@Measurement(iterations = 5)
public class ClassFilterBenchmark {

    private ClassFilter classFilter;

    @Setup
    public void prepare() {
        classFilter = new ClassFilter();
        classFilter.addPackages("foo");
    }

    @Benchmark
    public void litter(Blackhole blackhole) {
        blackhole.consume(classFilter.isListed_litter("foo.bar"));
    }

    @Benchmark
    public void noLitter(Blackhole blackhole) {
        blackhole.consume(classFilter.isListed_noLitter("foo.bar"));
    }
}
```

```

Benchmark                      Mode  Cnt   Score   Error  Units
ClassFilterBenchmark.litter    avgt    5  30.205 ± 7.995  ns/op
ClassFilterBenchmark.noLitter  avgt    5   6.386 ± 0.091  ns/op
```

Reply:
I added the gc profiler

```
Benchmark                                                     Mode  Cnt     Score     Error   Units
ClassFilterBenchmark.litter                                   avgt    5    29.432 ±   1.990   ns/op
ClassFilterBenchmark.litter:·gc.alloc.rate                    avgt    5  1555.331 ± 103.111  MB/sec
ClassFilterBenchmark.litter:·gc.alloc.rate.norm               avgt    5    48.000 ±   0.001    B/op
ClassFilterBenchmark.litter:·gc.churn.PS_Eden_Space           avgt    5  1550.413 ± 577.469  MB/sec
ClassFilterBenchmark.litter:·gc.churn.PS_Eden_Space.norm      avgt    5    47.876 ±  18.812    B/op
ClassFilterBenchmark.litter:·gc.churn.PS_Survivor_Space       avgt    5     0.181 ±   0.231  MB/sec
ClassFilterBenchmark.litter:·gc.churn.PS_Survivor_Space.norm  avgt    5     0.006 ±   0.007    B/op
ClassFilterBenchmark.litter:·gc.count                         avgt    5    21.000            counts
ClassFilterBenchmark.litter:·gc.time                          avgt    5    43.000                ms
ClassFilterBenchmark.noLitter                                 avgt    5     6.365 ±   0.122   ns/op
ClassFilterBenchmark.noLitter:·gc.alloc.rate                  avgt    5     0.001 ±   0.001  MB/sec
ClassFilterBenchmark.noLitter:·gc.alloc.rate.norm             avgt    5    ≈ 10⁻⁵              B/op
ClassFilterBenchmark.noLitter:·gc.count                       avgt    5       ≈ 0            counts
```